sql/core/src/test/scala/org/apache/spark/sql/DatasetCacheSuite.scala

```diff
@@ -134,6 +134,19 @@ class DatasetCacheSuite extends QueryTest
     assert(df.storageLevel == StorageLevel.NONE)
   }
 
+  test("SPARK-46992 collect before persisting") {
+    val ds = (1 to 4).toDF("id").sort("id").sample(0.4, 123)
+    // this check will call ds.collect() first
+    assert(Array(Row(1), Row(4)).sameElements(ds.collect()))
+    // and then cache it
+    ds.cache()
+    // Make sure, the Dataset is indeed cached.
+    assertCached(ds)
+    // Make sure the result of count() is consistent with collect()
```

Comment:
there is no need to the comments - they are saying the same thing what the code is doing